### ISSUE TITLE
SL Review: Delete button for submitted applications

### ISSUE THREAD

**Issue by HasanSifat** [bcgov/digital-journeys#609]
**User Story**
As an ADM who has already submitted an SL Review form. I might want to go back to the submitted form and delete it instead of editing it. 

**Tasks**

- [ ] Add delete button beside edit for all re-submit status applications

**Acceptance Criteria**
When the delete button has been added

**Comment by HasanSifat** [comment]
![image.png](https://images.zenhubusercontent.com/61e5b78d9a5a1eb2ba7a92c3/0a297988-24f8-43a2-b5a3-c40f4a2dbb1f)

**Comment by iman-jamali-fw** [comment]
@HasanSifat Does delete include deletion from ODS? And is this only for SL Review submission? 
In that case, @T854236 can we have a `delete` endpoint for the SL Review endpoint if confirmed by the PSA team?

**Comment by HasanSifat** [comment]
@iman-jamali-fw yes that includes deletion from the ODS endpoint as well.

**Comment by iman-jamali-fw** [comment]
@MeghanStothers just want to confirm that this delete application feature is only for SL Review. You don't want Telework forms to get deleted through the applications tab in the TOOL?

**Comment by MeghanStothers** [comment]
@iman-jamali-fw That's correct - we don't want Telework forms to get deleted before the expiry date

**Comment by iman-jamali-fw** [comment]
@T854236 I removed an application by id `6770` in ODS. Can you please confirm if it was deleted? Thank you

**Comment by T854236** [comment]
@iman-jamali-fw 

It is still there.  According to logs, the last call I see it is a POST with ID 6770.  Was that when you made the call?

2022-10-24T14:33:26 | Info | ~/api.rsc | Retrieving insert row from query: SELECT * FROM [ods_datamart_telework].[app_slreview_info] WHERE [applicationId] = ?;1=6770(string)

2022-10-24T14:33:26 | Datamart_Telework_app_slreview_info | telework | POST | https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info

**Comment by iman-jamali-fw** [comment]
@T854236  Yes, the post should have been my initial call. right after, I made a delete request, my logs show no error in the delete call.

**Comment by HasanSifat** [comment]
@iman-jamali-fw just made a new submission for SL review form in DEV. Don't see any delete button beside the submission in submitted forms tab

**Comment by iman-jamali-fw** [comment]
@HasanSifat This is what I see in DEV. 
![Screen Shot 2022-10-24 at 1.52.32 PM.png](https://images.zenhubusercontent.com/61ddc48547f476ed3f35ca28/4744f90f-46ad-4880-a247-74f7ff9c0cb5)

**Comment by T854236** [comment]
@iman-jamali-fw unfortunately, I don't see the call coming to the API server.  I see 2 additional POST with IDs 6771 and 6772.

**Comment by HasanSifat** [comment]
This is what i see

![image.png](https://images.zenhubusercontent.com/61e5b78d9a5a1eb2ba7a92c3/9d3a0b44-9ba5-499d-a86d-a6286dc29358)

**Comment by iman-jamali-fw** [comment]
@HasanSifat  Oh OK I think yours hasn't picked up the latest changes in DEV. Can you do a hard refresh and try again, please?

**Comment by HasanSifat** [comment]
Tested successfully in DEV- 
![image.png](https://images.zenhubusercontent.com/61e5b78d9a5a1eb2ba7a92c3/a5b9e528-b794-4dd3-8c06-8319a5e114fc)

**Comment by iman-jamali-fw** [comment]
@T854236 
I did another call to delete applicationId `6774`. Can you check to see if it is deleted?

Also, I tried updating an application in ODS and got a `403` error. Can you please check the logs and share more details on why it's failing?
`*__checkpoint ⇢ 403 from PUT https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info/6775 [DefaultWebClient]`

**Comment by T854236** [comment]
@iman-jamali-fw On 6774, it is still there.

When passing the ID to the endpoint, please use this format  https://analytics-
testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info(applicationId)

**Comment by iman-jamali-fw** [comment]
@T854236  OK so the problem is the put/delete request. Can you make an example of how a delete/put with `applicationId` works? I don't have access to BCGOV VPN so cannot see the endpoint details. 

Do you mean something like this?
`https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info(6775)`

**Comment by T854236** [comment]
@iman-jamali-fw 

Yes correct on your example.  Example below for DELETE

Request:
curl --header "x-cdata-authtoken: MY_AUTH_TOKEN" -X DELETE "https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info('applicationId_1')"
Response:
Empty response with a "204 No Content" HTTP status line.

**Comment by T854236** [comment]
@iman-jamali-fw

We see 2 calls and both are POST for some reason.

2022-10-25T09:41:17 | Datamart_Telework_app_slreview_info | telework | POST | https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info
-- | -- | -- | -- | --

Retrieving insert row from query: SELECT * FROM [ods_datamart_telework].[app_slreview_info] WHERE [applicationId] = ?;1=6788(string)

2022-10-25T09:39:33 | Datamart_Telework_app_slreview_info | telework | POST | https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info
-- | -- | -- | -- | --


Retrieving insert row from query: SELECT * FROM [ods_datamart_telework].[app_slreview_info] WHERE [applicationId] = ?;1=6787(string)

**Comment by iman-jamali-fw** [comment]
@T854236 What I understand from your example is that for put/delete I need to wrap `applicatoinId` in a pair of `'`. for example if applicationId is 700, it'll be `('700')`? Currently, I send it as `(700)`. Is that the problem?

**Comment by T854236** [comment]
@iman-jamali-fw 

It can be in quotes or not.  I ran sample with SoapUI with a DELETE and below is what I see on the server.  Record 6788 has been deleted.


2022-10-25T12:35:57 | Datamart_Telework_app_slreview_info | telework_tes | DELETE | https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info('6788')
-- | -- | -- | -- | --



2022-10-25T12:35:57 | Info | ~/api.rsc | Executing the following query: DELETE FROM [ods_datamart_telework].[app_slreview_info] WHERE [ods_datamart_telework].[app_slreview_info].[applicationId] = '6788'
-- | -- | -- | --

**Comment by iman-jamali-fw** [comment]
Interesting! The way it's currently implemented is without quotes. Can you try `https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info(6787)` and see if it deletes? If not, I'll change and deploy. I don't have the VPN so I cannot test unless deploying the changes to DEV 😭

**Comment by T854236** [comment]
@iman-jamali-fw Can you try with 6789?  I already deleted 6787:).

**Comment by iman-jamali-fw** [comment]
@T854236  created a new application with id `6793`. Please try this one.

**Comment by T854236** [comment]
@iman-jamali-fw ok I see 6793 in ODS.  Please try the DELETE.

**Comment by iman-jamali-fw** [comment]
@T854236 Can you try to delete with `SoapUI` and without quotes? I want to make sure deleting without quotes works and that the problem is on my side. `https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info(6793)`

**Comment by T854236** [comment]
@iman-jamali-fw yes confirm it works.  The tests you made earlier all came in as POST instead of DELETE.  That is the only difference.


2022-10-25T12:51:14 | Info | ~/api.rsc | Executing the following query: DELETE FROM [ods_datamart_telework].[app_slreview_info] WHERE [ods_datamart_telework].[app_slreview_info].[applicationId] = '6793'
-- | -- | -- | --


2022-10-25T12:51:14 | Datamart_Telework_app_slreview_info | telework_tes | DELETE | https://analytics-testapi.psa.gov.bc.ca/apiserver/api.rsc/Datamart_Telework_app_slreview_info(6793)
-- | -- | -- | -- | --

**Comment by iman-jamali-fw** [comment]
@T854236 Thanks for the confirmation. I'll check the requests and will come back to you.

**Comment by iman-jamali-fw** [comment]
@T854236 I found a laptop that can connect to BC VPN and did a local test, turned out the `delete` returns `401` authentication error. I can get the application with my token but it doesn't work when I try to delete it.

cc @simensma-fresh

**Comment by T854236** [comment]
@iman-jamali-fw Oh that is a different error.  I made a modification to the user account.  Please try again.

**Comment by iman-jamali-fw** [comment]
@T854236 Just deleted one with id `6795`.

**Comment by T854236** [comment]
@iman-jamali-fw Confirmed it works.

**Comment by iman-jamali-fw** [comment]
Great! Thanks.  @T854236 

For PUT, I'll do an update and will let you know.

**Comment by iman-jamali-fw** [comment]
@T854236 There was an issue in the update application logic that sent update and post at the same time, I fixed that and just updated an application with id `6800`. Can you please check if it is updated on your end? I got no error.

**Comment by T854236** [comment]
@iman-jamali-fw 
Yes I think the PUT works also.

Sample log for one of them;


2022-10-25T17:19:13 | Info | ~/api.rsc | Executing the following query: UPDATE [ods_datamart_telework].[app_slreview_info] SET [esl_leadingLargeTeams] = ?, [essk_procurement] = ?, [ldl_communication] = ?, [ldl_valuingDiversityAndInclusion] = ?, [esl_conflictManagement] = ?, [lda_innovation] = ?, [esa_serviceOutcomesForCitizensAndClients] = ?, [ldsk_digitalInformationAndTechnology] = ?, [essk_legislationRegulationsAndPolicy] = ?, [essk_negotiations] = ?, [ldl_collaboration] = ?, [branch] = ?, [ldsk_riskManagement] = ?, [division] = ?, [HadAConversationWithTheSeniorLeaderAboutTheContentInThisForm] = ?, [essk_programDeliveryOperations] = ?, [essk_indigenousRelationsAndReconciliation] = ?, [esl_promotingEmpowerment] = ?, [ldl_leadingVirtually] = ?, [expertiseAndSkillsComments] = ?, [essk_other] = ?, [ldsk_negotiations] = ?, [ldl_relationshipBuilding] = ?, [essk_projectManagement] = ?, [essk_riskManagement] = ?, [keyAccomplishment1] = ?, [lda_solvingProblemsCreatively] = ?, [lda_serviceOutcomesForCitizensAndClients] = ?, [ValidatedThisSeniorLeaderReadinessToCompleteForAnExecutiveRole] = ?, [ldsk_intergovernmentalRelations] = ?, [lda_visionAndGoalSetting] = ?, [ldl_executivePresence] = ?, [LearningAndDevelopmentComments] = ?, [ldsk_programDeliveryOperations] = ?, [esa_designingStrategyAndStructure] = ?, [esa_creatingAndSupportingChange] = ?, [esa_visionAndGoalSetting] = ?, [admSupervisorName] = ?, [whenIsReady] = ?, [ministry] = ?, [esa_innovation] = ?, [employeeName] = ?, [lda_designingStrategyAndStructure] = ?, [ldsk_financialManagement] = ?, [employeeNumber] = ?, [ldsk_labourRelations] = ?, [doYouWantToAddAnotherResponsibility2] = ?, [esl_leadingVirtually] = ?, [ldl_promotingEmpowerment] = ?, [ldl_leadingLargeTeams] = ?, [essk_ethics] = ?, [ldsk_ethics] = ?, [ldsk_indigenousRelationsAndReconciliation] = ?, [applicationStatus] = ?, [employeePositionTitle] = ?, [summaryRecommendations] = ?, [employeeWorkCity] = ?, [ldsk_humanResources] = ?, [ldsk_procurement] = ?, [admSupervisorEmail] = ?, [esl_executivePresence] = ?, [ldsk_legislationRegulationsAndPolicy] = ?, [esl_collaboration] = ?, [keyResponsibility1] = ?, [applicationId] = ?, [esl_relationshipBuilding] = ?, [essk_cabinetAndTreasuryBoardSubmissions] = ?, [ldsk_projectManagement] = ?, [esl_communication] = ?, [ldsk_cabinetAndTreasuryBoardSubmissions] = ?, [essk_labourRelations] = ?, [essk_intergovernmentalRelations] = ?, [ldsk_other] = ?, [essk_humanResources] = ?, [lda_creatingAndSupportingChange] = ?, [esa_solvingProblemsCreatively] = ?, [essk_financialManagement] = ?, [ldl_conflictManagement] = ?, [essk_digitalInformationAndTechnology] = ?, [doYouWantToAddAnotherAccomplishment2] = ?, [essk_executiveBriefing] = ?, [esl_valuingDiversityAndInclusion] = ?, [ldsk_executiveBriefing] = ? WHERE [ods_datamart_telework].[app_slreview_info].[applicationId] = '6800';1=false(bool);2=false(bool);3=false(bool);4=false(bool);5=false(bool);6=false(bool);7=false(bool);8=false(bool);9=false(bool);10=false(bool);11=false(bool);12=Forecasting (WDA)(string);13=false(bool);14=Labour Market Analytics, Forecasting, and Information Branch(string);15=true(bool);16=false(bool);17=false(bool);18=false(bool);19=false(bool);20=null(string);21=false(bool);22=false(bool);23=false(bool);24=false(bool);25=false(bool);26=sdf(string);27=false(bool);28=false(bool);29=true(bool);30=false(bool);31=false(bool);32=false(bool);33=null(string);34=false(bool);35=false(bool);36=false(string);37=false(bool);38=sdf(string);39=now(string);40=Advanced Education and Skills Training(string);41=false(bool);42=Hasan,Sazid(string);43=false(bool);44=false(bool);45=169848(string);46=false(bool);47=null(bool);48=false(bool);49=false(bool);50=false(bool);51=false(bool);52=false(bool);53=false(bool);54=Submitted(string);55=Economist(string);56=null(string);57=Vancouver(string);58=false(bool);59=false(bool);60=sdf@sdf.sd(string);61=false(bool);62=false(bool);63=false(string);64=sdf(string);65=6800(string);66=false(bool);67=false(bool);68=false(bool);69=false(bool);70=false(bool);71=false(string);72=false(bool);73=false(bool);74=false(bool);75=false(bool);76=false(bool);77=false(bool);78=false(bool);79=false(bool);80=null(bool);81=false(bool);82=false(bool);83=false(bool)
-- | -- | -- | --

**Comment by iman-jamali-fw** [comment]
Woo-hoo! 🎉 Thank you so much @T854236

**Comment by HasanSifat** [comment]
Ready to move to TEST!

**Comment by JodiLittle** [comment]
Approved

**Comment by HasanSifat** [comment]
tested successfully in TEST
